src/flamenco/vm/syscall/fd_vm_syscall_util.c

```diff
@@ -380,97 +380,97 @@ fd_vm_syscall_sol_memcpy( /**/            void *  _vm,
+        FD_VM_ERR_FOR_LOG_EBPF( vm, FD_VM_ERR_EBPF_ACCESS_VIOLATION );
+        return FD_VM_ERR_SIGSEGV;
       }
 
+      /* There should be no bytes eligible for copying if the input region is not writable. */
+      dst_bytes_rem_in_cur_region = fd_ulong_sat_sub( vm->input_mem_regions[ dst_region_idx ].region_sz, (dst_offset - vm->input_mem_regions[ dst_region_idx ].vaddr_offset) );
     } else {
-      dst_haddr = (uchar *)FD_VM_MEM_SLICE_HADDR_ST( vm, dst_vaddr, FD_VM_ALIGN_RUST_U8, sz );
+      dst_haddr = (uchar*)FD_VM_MEM_HADDR_ST_FAST( vm, dst_vaddr );
+      dst_bytes_rem_in_cur_region = fd_ulong_min( sz, fd_ulong_sat_sub( vm->region_st_sz[ dst_region ], dst_offset ) );
     }
 
-    ulong src_region              = src_vaddr >> 32;
-    ulong src_offset              = src_vaddr & 0xffffffffUL;
-    ulong src_region_idx          = 0UL;
-    ulong src_bytes_in_cur_region = sz;
-    uchar * src_haddr             = NULL;
-    if( src_region==4UL ) {
-      src_region_idx          = fd_vm_get_input_mem_region_idx( vm, src_offset );
-      src_haddr               = (uchar*)(vm->input_mem_regions[ src_region_idx ].haddr + src_offset - vm->input_mem_regions[ src_region_idx ].vaddr_offset);
-      src_bytes_in_cur_region = fd_ulong_min( sz, fd_ulong_sat_sub( vm->input_mem_regions[ src_region_idx ].region_sz, 
-                                                                    ((ulong)src_haddr - vm->input_mem_regions[ src_region_idx ].haddr) ) );
-      if( FD_UNLIKELY( src_region_idx+1UL==vm->input_mem_regions_cnt && src_bytes_in_cur_region<sz ) ) {
-        *_ret = 1;
-        return FD_VM_ERR_ABORT;
-      }
-
+    ulong   src_region                  = fd_ulong_min( src_vaddr >> 32, 5UL );
+    uchar   src_is_input_mem_region     = ( src_region==4UL );
+    ulong   src_offset                  = src_vaddr & 0xffffffffUL;
+    ulong   src_region_idx              = 0UL;
+    ulong   src_bytes_rem_in_cur_region;
+    uchar * src_haddr;
+    if( src_is_input_mem_region ) {
+      FD_VM_MEM_HADDR_AND_REGION_IDX_FROM_INPUT_REGION_UNCHECKED( vm, src_offset, src_region_idx, src_haddr );
+      src_bytes_rem_in_cur_region = fd_ulong_sat_sub( vm->input_mem_regions[ src_region_idx ].region_sz, (src_offset - vm->input_mem_regions[ src_region_idx ].vaddr_offset) );
     } else {
-      src_haddr           = (uchar *)FD_VM_MEM_SLICE_HADDR_LD( vm, src_vaddr, FD_VM_ALIGN_RUST_U8, sz );
+      src_haddr = (uchar*)FD_VM_MEM_HADDR_LD_FAST( vm, src_vaddr );
+      src_bytes_rem_in_cur_region = fd_ulong_min( sz, fd_ulong_sat_sub( vm->region_ld_sz[ src_region ], src_offset ) );
     }
 
-    /* Do a normal memcpy if regions do not overlap */
-    if( FD_LIKELY( src_bytes_in_cur_region==dst_bytes_in_cur_region && src_bytes_in_cur_region==sz ) ) {
+    /* Short circuit: if the number of copyable bytes stays within all memory regions, just memcpy and return.
+       This is a majority case in mainnet. */
```

Comment:
kind of a nit, but this almost always happens with malicious txns, you can make that more clear